EIPS/eip-2014.md

```diff
@@ -0,0 +1,94 @@
+---
+eip: 2014
+title: Extended State Oracle
+author: Alex Beregszaszi (@axic)
+discussions-to: <URL>
+status: Draft
+type: Standards Track
+category: Core
+created: 2019-05-10
+requires: 140
+---
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+
+## Abstract
+
+Introduce a new system contract with an extensible interface following the [Contract ABI Encoding] to access extended data sets, such as chain identifiers, block hashes, etc.
+
+This allows Ethereum contract languages to interact with this contract as if it were a regular contract and not needing any language support.
+
+## Motivation
+
+Over the past couple of years several proposals were made to extend the EVM with more data. Some examples include extended access to block hashes ([EIP-210]) and chain identifiers ([EIP-1344]).
+
+Adding them as EVM opcodes seems to be using the scarce opcode space for relatively less frequently used features, while adding them as precompiles is perceived as more complicated due to an interface
+needs to be defined and agreed on for every case.
+
+This proposal tries to solve both issues with defining an extensible standard interface.
+
+## Specification
+
+A new system contract ("precompile") is introduced at address `0x0000000000000000000000000000000000000009` called ESO (Extended State Oracle).
+
+It can be queried using `CALL` or `STATICCALL` and follows the [Contract ABI Encoding] for the inputs and outputs. Using elementary types in the ABI encoding is encouraged to keep complexity low.
+
+In the future it could be possible to extend ESO to have a state and accept transactions from a system address to store the passed data -- similarly to what [EIP-210] proposed.
+
+Proposals wanting to introduce more data to the state, which is not part of blocks or transactions, should aim to extend the ESO.
+
+At this time it is not proposed to make the ESO into a contract existing in the state, but to include it as a precompile and leave the implementation details to the client.
+In the future if it is sufficiently extended and a need arises to have a state, it would make sense to move it from being a precompile and have actual code.
+
+### Chain identifier
+
+Initially, a feature to read the current chain identifier is introduced: `getCurrentChainId()` returns the current chain identifier as a `uint64` encoded value.
```

Comment:
Should return `uint256`. In the discussions for chain ID, this value currently doesn't not exceed `uint64` but at a future point in time it could make sense to use a 32-byte hash reference, that is the computation of a few things (genesis block, current EIPs, etc.)